### PR TITLE
[NFC] Adopt TypeBase-isms for GenericSignature

### PR DESCRIPTION
**Pull request by CodaFi**
Structurally prevent a number of common anti-patterns involving generic
signatures by separating the interface into GenericSignature and the
implementation into GenericSignatureBase.  In particular, this allows
the comparison operators to be deleted which forces callers to
canonicalize the signature or ask to compare pointers explicitly.

### CODE REVIEW

**Comment by CodaFi** [comment]
apple/swift-lldb#2022
@swift-ci please test

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -388,7 +388,8 @@ ValueDecl *RequirementFailure::getDeclRef() const {
       do {
         if (auto *parent = DC->getAsDecl()) {
           if (auto *GC = parent->getAsGenericContext()) {
-            if (GC->getGenericSignature() != Signature)
+            // FIXME: Is this intending an exact match?
+            if (GC->getGenericSignature().getPointer() != Signature.getPointer())
```
Comment (CodaFi) [inline lib/Sema/CSDiagnostics.cpp]: @xedin Should this canonicalize the generic signatures, or are we looking for an exact match here?

**Review by DougGregor — APPROVED** [review]
Yes please!

**Comment by jrose-apple** [comment]
Nitpick: because there are no subclasses, can you call it GenericSignatureImpl or signature?

**Comment by jrose-apple** [comment]
Alternately, the indirect one could be GenericSignatureRef, like ProtocolConformanceRef.

**Comment by CodaFi** [comment]
I lean towards `GenericSignatureImpl` to keep noise in the diff down.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/swift-lldb#2022
@swift-ci please smoke test

**Comment by CodaFi** [comment]
⛵️

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -541,7 +541,7 @@ class alignas(1 << TypeAlignInBits) TypeBase {
   
   /// allowsOwnership() - Are variables of this type permitted to have
   /// ownership attributes?
-  bool allowsOwnership(GenericSignature *sig=nullptr);
+  bool allowsOwnership(GenericSignatureImpl *sig = nullptr);
```
Comment (slavapestov) [inline include/swift/AST/Types.h]: Why isn't this a GenericSignature?

File: lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -4376,15 +4377,15 @@ ASTContext::getOverrideGenericSignature(const ValueDecl *base,
     return nullptr;
   }
 
-  if (derivedClass->getGenericSignature() == nullptr &&
+  if (derivedClass->getGenericSignature().isNull() &&
```
Comment (slavapestov) [inline lib/AST/ASTContext.cpp]: Is there no implicit conversion to bool?

File: lib/AST/ASTContext.cpp [hunk 4]
```diff
@@ -4440,13 +4441,13 @@ ASTContext::getOverrideGenericSignature(const ValueDecl *base,
     }
   }
 
-  auto *genericSig = evaluateOrDefault(
+  auto genericSig = evaluateOrDefault(
       ctx.evaluator,
       AbstractGenericSignatureRequest{
-        derivedClass->getGenericSignature(),
+        derivedClass->getGenericSignature().getPointer(),
```
Comment (slavapestov) [inline lib/AST/ASTContext.cpp]: This unwrapping should not be necessary

File: lib/AST/ASTDemangler.cpp [hunk 5]
```diff
@@ -446,7 +446,7 @@ Type ASTBuilder::createImplFunctionType(
     ArrayRef<Demangle::ImplFunctionResult<Type>> results,
     Optional<Demangle::ImplFunctionResult<Type>> errorResult,
     ImplFunctionTypeFlags flags) {
-  GenericSignature *genericSig = nullptr;
+  GenericSignature genericSig = GenericSignature();
```
Comment (slavapestov) [inline lib/AST/ASTDemangler.cpp]: It should be default initialized

File: lib/AST/GenericEnvironment.cpp [hunk 6]
```diff
@@ -215,7 +215,7 @@ GenericEnvironment::mapConformanceRefIntoContext(
                                      ProtocolConformanceRef conformance) const {
   auto contextConformance = conformance.subst(conformingInterfaceType,
     QueryInterfaceTypeSubstitutions(this),
-    LookUpConformanceInSignature(*getGenericSignature()));
+    LookUpConformanceInSignature(getGenericSignature().getPointer()));
```
Comment (slavapestov) [inline lib/AST/GenericEnvironment.cpp]: More interfaces to clean up

**Comment by slavapestov** [comment]
I think we can go further here and make GenericSignature like SubstitutionMap. Instead of forwarding method calls from GenericSignature to GenericSignatureImpl with operator-> we can implement the methods on GenericSignature itself. This way we can also implement meaningful behavior for the empty value -- just treat it as a signature with no parameters and no requirements. This would in turn simplify away lots of null checks all over the place.

**Pull request merged**